lib/dialects/abstract/query-generator.js

```diff
@@ -1578,128 +1273,294 @@ const QueryGenerator = {
+        if (verbatim === true) {
+          prefix = attr;
+        } else {
+          prefix = `${this.quoteIdentifier(includeAs.internalAs)}.${this.quoteIdentifier(attr)}`;
+        }
+        return `${prefix} AS ${this.quoteIdentifier(`${includeAs.externalAs}.${attrAs}`, true)}`;
+      });
+      if (include.subQuery && topLevelInfo.subQuery) {
+        for (let attr of includeAttributes) {
+          attributes.subQuery.push(attr);
+        }
+      } else {
+        for (let attr of includeAttributes) {
+          attributes.main.push(attr);
+        }
       }
+    }
+
+    //through
+    if (include.through) {
+      joinQuery = this.generateThroughJoin(include, includeAs, parentTableName.internalAs, topLevelInfo);
     } else {
-      var sql = this.quote(typeof options.order === 'string' ? new Utils.Literal(options.order) : options.order, model);
-      if (subQuery) {
-        subQueryOrder.push(sql);
+
+
+      if (topLevelInfo.subQuery && include.subQueryFilter) {
+        const associationWhere = {};
+
+        associationWhere[association.identifierField] = {
+          $raw: `${this.quoteTable(parentTableName.internalAs)}.${this.quoteIdentifier(association.source.primaryKeyField)}`
+        };
+
+        if (!topLevelInfo.options.where) {
+          topLevelInfo.options.where = {};
+        }
+
+        // Creating the as-is where for the subQuery, checks that the required association exists
+        const $query = this.selectQuery(include.model.getTableName(), {
+          attributes: [association.identifierField],
+          where: {
+            $and: [
+              associationWhere,
+              include.where || {}
+            ]
+          },
+          limit: 1
+        }, include.model);
+
+        const subQueryWhere = this.sequelize.asIs([
+          '(',
+          $query.replace(/\;$/, ''),
+          ')',
+          'IS NOT NULL'
+        ].join(' '));
+
+        if (Utils._.isPlainObject(topLevelInfo.options.where)) {
+          topLevelInfo.options.where['__' + includeAs] = subQueryWhere;
+        } else {
+          topLevelInfo.options.where = { $and: [topLevelInfo.options.where, subQueryWhere] };
+        }
       }
-      mainQueryOrder.push(sql);
+      joinQuery = this.generateJoin(include, topLevelInfo);
     }
 
-    return {mainQueryOrder, subQueryOrder};
-  },
+    // handle possible new attributes created in join
+    if (joinQuery.attributes.main.length > 0) {
+      attributes.main = attributes.main.concat(joinQuery.attributes.main);
+    }
 
-  selectFromTableFragment(options, model, attributes, tables, mainTableAs) {
-    let fragment = 'SELECT ' + attributes.join(', ') + ' FROM ' + tables;
+    if (joinQuery.attributes.subQuery.length > 0) {
+      attributes.subQuery = attributes.subQuery.concat(joinQuery.attributes.subQuery);
+    }
 
-    if(mainTableAs) {
-      fragment += ' AS ' + mainTableAs;
+    if (include.include) {
+
+      for (const childInclude of include.include) {
```

Comment:
Minor nitpick, you seem to add an extra newline after each `{` which is not how we generally do it in the rest of the codebase (goes for a large part of the diff)